# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1770081187 -39600
# Parent  813839792068096cc635c8288705e86e4a189d08
Feature: Add web search to context menu in compose window.

diff --git a/mail/components/compose/content/MsgComposeCommands.js b/mail/components/compose/content/MsgComposeCommands.js
--- a/mail/components/compose/content/MsgComposeCommands.js
+++ b/mail/components/compose/content/MsgComposeCommands.js
@@ -49,8 +49,12 @@ var { ExtensionParent } = ChromeUtils.im
 var { PlacesUtils } = ChromeUtils.importESModule(
   "resource://gre/modules/PlacesUtils.sys.mjs"
 );
 
+var { openWebSearch } = ChromeUtils.importESModule(
+  "resource:///modules/LinkHelper.sys.mjs"
+);
+
 ChromeUtils.defineESModuleGetters(this, {
   AttachmentInfo: "resource:///modules/AttachmentInfo.sys.mjs",
   BondOpenPGP: "chrome://openpgp/content/BondOpenPGP.sys.mjs",
   EnigmailKeyRing: "chrome://openpgp/content/modules/keyRing.sys.mjs",
@@ -101,8 +105,12 @@ const lazy = {};
 ChromeUtils.defineESModuleGetters(lazy, {
   MailStringUtils: "resource:///modules/MailStringUtils.sys.mjs",
 });
 
+const messengerBundle = Services.strings.createBundle(
+  "chrome://messenger/locale/messenger.properties"
+);
+
 /**
  * Global message window object. This is used by mail-offline.js and therefore
  * should not be renamed. We need to avoid doing this kind of cross file global
  * stuff in the future and instead pass this object as parameter when needed by
@@ -1981,13 +1989,65 @@ function openLinkInBrowser() {
     .getService(Ci.nsIExternalProtocolService)
     .loadURI(lastUriHovered, principal);
 }
 
+function composeOpenWebSearch() {
+  const selection = GetCurrentEditor().selection;
+  const text = selection
+    .getRangeAt(0)
+    .startContainer.textContent.substring(
+      selection.getRangeAt(0).startOffset,
+      selection.getRangeAt(0).endOffset
+    )
+    .trim();
+  openWebSearch(text);
+}
+
 function msgComposeContextOnShowing(event) {
   if (event.target.id != "msgComposeContext") {
     return;
   }
 
+  // Web search.
+  const searchTheWeb = document.getElementById("searchTheWeb");
+  const searchTheWebSeparator = document.getElementById(
+    "searchTheWebSeparator"
+  );
+  const selection = GetCurrentEditor().selection;
+  if (
+    selection.isCollapsed ||
+    selection.rangeCount != 1 ||
+    selection.getRangeAt(0).startContainer !=
+      selection.getRangeAt(0).endContainer ||
+    selection.getRangeAt(0).startContainer.nodeType != Node.TEXT_NODE
+  ) {
+    searchTheWeb.hidden = true;
+    searchTheWebSeparator.hidden = true;
+  } else {
+    searchTheWeb.hidden = false;
+    searchTheWebSeparator.hidden = false;
+    const text = selection
+      .getRangeAt(0)
+      .startContainer.textContent.substring(
+        selection.getRangeAt(0).startOffset,
+        selection.getRangeAt(0).endOffset
+      )
+      .trim();
+    let key = "openSearch.label";
+    let abbrSelection;
+    if (text.length > 15) {
+      key += ".truncated";
+      abbrSelection = text.slice(0, 15);
+    } else {
+      abbrSelection = text;
+    }
+
+    searchTheWeb.label = messengerBundle.formatStringFromName(key, [
+      Services.search.defaultEngine.name,
+      abbrSelection,
+    ]);
+  }
+
   // gSpellChecker handles all spell checking related to the context menu,
   // except whether or not spell checking is enabled. We need the editor's
   // spell checker for that.
   gSpellChecker.initFromRemote(
diff --git a/mail/components/compose/content/messengercompose.xhtml b/mail/components/compose/content/messengercompose.xhtml
--- a/mail/components/compose/content/messengercompose.xhtml
+++ b/mail/components/compose/content/messengercompose.xhtml
@@ -655,8 +655,12 @@
   <menuseparator/>
   <menuitem id="openLinkInBrowser" label="&openLinkInBrowser.label;" oncommand="openLinkInBrowser();"/>
   <menuitem data-l10n-id="text-action-select-all" command="cmd_selectAll"/>
 
+  <!-- Search -->
+  <menuseparator id="searchTheWebSeparator"/>
+  <menuitem id="searchTheWeb" oncommand="composeOpenWebSearch()"/>
+
   <!-- Spellchecking general menu items (enable, add dictionaries...) -->
   <menuseparator id="spellCheckSeparator"/>
   <menuitem id="spellCheckEnable"
             data-l10n-id="text-action-spell-check-toggle"
